lib/crypto/aes.toit

```diff
@@ -2,92 +2,155 @@
 
-  The $key must be 32 secret bytes and the $initialization_vector must be 16 bytes.
+  The $key must be 16, 24 or 32 secret bytes and the 
+  $initialization_vector must be 16 bytes.
   */
   constructor.decryptor key/ByteArray initialization_vector/ByteArray:
-    return AesCbc.common_ key initialization_vector false
+    super.from_subclass key initialization_vector false
 
-  constructor.common_ key/ByteArray initialization_vector/ByteArray encrypt/bool:
-    aes_ = aes_cbc_init_ resource_freeing_module_ key initialization_vector encrypt
-    add_finalizer this:: this.close
+  /** See $super. */
+  crypt_ input/ByteArray encrypt/bool -> ByteArray:
+    from := 0
+    to := input.size
+    return aes_cbc_crypt_ aes_ input from to encrypt
+
+/**
+Advanced Encryption Standard Electronic codebook (AES-ECB).
+
+This implementation uses hardware accelerated primitives.
+
+AES-ECB state for encrypting and decrypting 
+
+To encrypt, construct an AES encryption state with $AesEcb.encryptor and
+  encrypt a block with $encrypt.
+
+To decrypt, construct and AES decryption state with $AesEcb.decryptor and
+  decrypt a block with $decrypt.
+
+Close the AES state with $close to release system resources.
+*/
+class AesEcb extends Aes:
 
   /**
-  Encrypts the given $cleartext.
+  Creates an AES-ECB state for encryption.
 
-  The size of the $cleartext must be a multiple of 16.
-  Returns a byte array with the encrypted data.
+  The $key must be either 16, 24 or 32 secret bytes.
   */
-  encrypt cleartext/ByteArray -> ByteArray:
-    return crypt_ cleartext true
+  constructor.encryptor key/ByteArray:
+    super.from_subclass key (ByteArray 0) true
 
   /**
-  Decrypts the given $ciphertext.
-  The size of the $ciphertext must be a multiple of 16.
-  Returns a byte array with the decrypted data.
+  Creates an AES-ECB state for decryption.
+
+  The $key must be either 16, 24 or 32 secret bytes.
   */
-  decrypt ciphertext/ByteArray -> ByteArray:
-    return crypt_ ciphertext false
+  constructor.decryptor key/ByteArray:
+    super.from_subclass key (ByteArray 0) false
 
+  /** See $super. */
   crypt_ input/ByteArray encrypt/bool -> ByteArray:
     from := 0
     to := input.size
-    if not aes_: throw "ALREADY_CLOSED"
-    result := aes_cbc_crypt_ aes_ input from to encrypt
-    return result
+    return aes_ecb_crypt_ aes_ input from to encrypt
 
-  /** Closes this encrypter and releases associated resources. */
-  close -> none:
-    aes_cbc_close_ aes_
-    aes_ = null
-    remove_finalizer this
 
-aes_cbc_init_ group key/ByteArray initialization_vector/ByteArray encrypt/bool:
-  #primitive.crypto.aes_cbc_init
+
+aes_init_ group key/ByteArray initialization_vector/ByteArray? encrypt/bool:
+  #primitive.crypto.aes_init
 
 aes_cbc_crypt_ aes input/ByteArray from/int to/int encrypt/bool:
   #primitive.crypto.aes_cbc_crypt
 
-aes_cbc_close_ aes:
-  #primitive.crypto.aes_cbc_close
+aes_ecb_crypt_ aes input/ByteArray from/int to/int encrypt/bool:
```

Comment:
Maybe we should drop the `from` and `to` arguments here (and for `aes_cbc_crypt_`) and just use slices? These are always called with `0` and `input.size`.

Reply:
The from and to arguments have been removed all the way down to the primitive. We now use the length of the Blob to test for size constraints.